### PR TITLE
feat(ci): Add .deb release to pipeline

### PR DESCRIPTION
**Pull request by IZOBRETATEL777**
Add .deb release
## Description
Added auto build of .deb packages for Debian-based Linux distributions.

## Motivation and Context
Support of Ubuntu, Linux Mint and other Deb-Linuxes
No need of manually compilation and etc.
Easy for beginners   

## How Has This Been Tested?

Ubuntu 20.04
Xubuntu 18.04
Debian 10 Buster (XFCE)
Debian 10 Buster (KDE)

**Not working on Ubuntu 16.04** due to OpenSSL 1.1.1 required.

## Type of changes

- [ ] Bug fix (changes which fix an issue)
- [x] New feature (changes which add functionality)
- [ ] Documentation (changes which modify the documentation only)
- [ ] Style change (changes which do not affect the meaning of the code: code formatting, etc.)
- [ ] Refactor (changes which affect the meaning of the code but neither fix a bug nor add a feature)
- [ ] Performance improve (changes which improve performance)
- [ ] Test (changes which add tests)
- [x] Build (Changes that affect the build system or external dependencies)
- [x] CI (changes to CI configuration files and scripts)
- [ ] Chore (changes which do not belong to any type above)
- [ ] Revert (revert previous changes)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/cpeditor/cpeditor/blob/master/CONTRIBUTING.md) document.
- [x] I have tested these changes locally, and this fixes the bug/the new feature behaves as the expectation.
- [x] The settings file in the old version can be used in the new version after this change.
- [x] These changes only fix a single bug/introduces a single feature. (Otherwise, open multiple Pull Requests instead, unless these bugs/features are closely related.)
- [x] The commit messages are clear and detailed. (Otherwise, use `git reset` and commit again, or use `git rebase -i` and `git commit --amend` to modify the commit messages.)
- [x] These changes don't remove an existing feature. (Otherwise, add an option to disable this feature instead, unless it's necessary to remove this feature.)
- [x] If there are changes of the text displayed in the UI, I have wrapped them in `tr()` or `QCoreApplication::translate()`.
- [ ] I have documented these changes in [CHANGELOG.md](https://github.com/cpeditor/cpeditor/blob/master/doc/CHANGELOG.md), or these changes are not notable.

@ouuan please check translations of cpeditor.desktop

### CODE REVIEW

File: .ci/DEB/cpeditor.desktop
Comment (coder3101) [inline .ci/DEB/cpeditor.desktop:12]: File extension not required.

It's better to have only one desktop file for Linux. Please modify the already existing file at ci/linux/cpeditor.desktop

File: .github/workflows/release_deb.yml
Comment (coder3101) [inline .github/workflows/release_deb.yml:67]: These mkdirs can be reduced by using -p flag and [path1, path2] bash script notations.

File: .github/workflows/release_deb.yml
Comment (coder3101) [inline .github/workflows/release_deb.yml:72]: Use the environment variable that points to Source location `${{ github.workspace }}/cpeditor/build/plugins`

File: .github/workflows/release_deb.yml
Comment (coder3101) [inline .github/workflows/release_deb.yml:76]: Is it required to move translations files?

File: .github/workflows/release_deb.yml
Comment (coder3101) [inline .github/workflows/release_deb.yml:102]: I think you can release this deb with the current existing release.yml. Just add a new Matrix variable and copy paste steps by adding condition with `if`. What's your opinion @ouuan ?

File: .ci/DEB/cpeditor.desktop
Comment (coder3101) [inline .ci/DEB/cpeditor.desktop:12]: Just put `cpeditor-icon` it knows the location to look for that file and the file extension.

**Comment by coder3101** [comment]
You can provide a link to your GitHub repository where you have tested this script and able to release the deb file. It convinces us that at time of release your script will run completely without errors.

**Comment by coder3101** [comment]
`master` -> `master`???

File: .ci/linux/cpeditor.desktop
Comment (ouuan) [inline .ci/linux/cpeditor.desktop:7]: Are you sure we should have a long name? "CP" comes from "Competitive Programming", but I think using "Competitive Programming Editor“ as a long name of CP Editor is weird.

File: .ci/linux/cpeditor.desktop
Comment (ouuan) [inline .ci/linux/cpeditor.desktop:11]: The comment here should be a short line instead of a long sentence. And the "CP Editor is" is redundant here.

File: .github/workflows/release_deb.yml
Comment (ouuan) [inline .github/workflows/release_deb.yml:6]: It should be in the Release workflow, instead of a new workflow. And the most important, it should be triggered on tags instead of when a release is published.

File: .ci/linux/cpeditor.desktop
Comment (ouuan) [inline .ci/linux/cpeditor.desktop:7]:
```suggestion
GenericName=Code editor for competitive programming
GenericName[ru]=Редактор кода для спортивного программирования
GenericName[zh_CN]=算法竞赛代码编辑器
```

**Comment by IZOBRETATEL777** [comment]
> Is there any reason to use "DEBIAN" everywhere? I think "Debian" is better.
As I know we should use "DEBIAN" (with uppercase letters) for name of main directory with service files.
image

**Comment by ouuan** [comment]
> > Is there any reason to use "DEBIAN" everywhere? I think "Debian" is better.
> 
> As I know we should use "DEBIAN" (with uppercase letters) for name of main directory with service files.
> image

If it is required, then it's OK.

File: .ci/DEB/cpeditor.desktop
Comment (IZOBRETATEL777) [reply .ci/DEB/cpeditor.desktop]: I think that extension (.png) is required:
image

File: .github/workflows/release_deb.yml
Comment (IZOBRETATEL777) [reply .github/workflows/release_deb.yml]: No, we can copy too. I'll fix it.

Comment (coder3101) [reply .github/workflows/release_deb.yml]: No I mean where is raw translation files used? Is it required by Qt to be present for translation to work?

File: .ci/DEB/cpeditor.desktop
Comment (coder3101) [reply .ci/DEB/cpeditor.desktop]: Read https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html

You need to only specify the name of the icon-file. The extension (png) is added automatically. It is not good idea to have an absolute path to icon file because this way when user changes its theme/style the icon will not change.

File: .ci/linux/cpeditor.desktop
Comment (coder3101) [inline .ci/linux/cpeditor.desktop:12]:
```suggestion
Icon=cpeditor
```

File: .github/workflows/release_deb.yml
Comment (coder3101) [inline .github/workflows/release_deb.yml:80]:
```suggestion
          cp ${{ github.workspace }}/.ci/linux/cpeditor.png ./icons/cpeditor.png
```

File: .github/workflows/release_deb.yml
Comment (ouuan) [inline .github/workflows/release_deb.yml:4]: Release on push? It should be in `release.yml`.

Comment (IZOBRETATEL777) [reply .github/workflows/release_deb.yml]: Yes, It was for debugging. Now all in the main Release workflow

**Comment by IZOBRETATEL777** [comment]
> You can provide a link to your GitHub repository where you have tested this script and able to release the deb file. It convinces us that at time of release your script will run completely without errors.

https://github.com/IZOBRETATEL777/test_pipeline

**Comment by IZOBRETATEL777** [comment]
> Is there any reason to use "DEBIAN" everywhere? I think "Debian" is better.

We can also move files from .ci/DEB to .ci/linux folder for better clarity

**Comment by IZOBRETATEL777** [comment]
Should I update the documentation and write about deb? Or it is problem for next PR?

**Comment by coder3101** [comment]
> Should I update the documentation and write about deb? Or it is problem for next PR?

Yes this way it will be a complete PR. You should open a PR to website repository. We will merge it just before we publish first Deb pakage. (Next release)

**Comment by coder3101** [comment]
> > Is there any reason to use "DEBIAN" everywhere? I think "Debian" is better.
> 
> We can also move files from .ci/DEB to .ci/linux folder for better clarity

I think .ci/DEB is okay. ci/linux is for something that's valid for all linux distro, like icons and desktop file.

**Comment by coder3101** [comment]
You should modify `doc/INSTALL` and `doc/INSTALL(Russian)` . While approving @ouuan will add same in Chinese Readme. Next after we merge this PR, you should open a PR to add it to website. We will merge it when releasing new update.

**Comment by coder3101** [comment]
You also need to add the SHA checksum of Deb to `release_checksum.yml` as well.

**Comment by coder3101** [comment]
Can you rename the debian package to cpeditor-X.Y.Z-amd64.deb. It's currently of format cpeditor_X.Y.Z_amd64.deb

**Comment by ouuan** [comment]
I'm wondering whether we should provide this in the download page of the website, and if we should, how? And what about the update checker :thinking:

**Comment by ouuan** [comment]
I tested the CI in my fork and tested the package with [debtap](https://aur.archlinux.org/packages/debtap) on Arch Linux. It works.

Now the problems are the documentation, download page and the update checker.

I'm wondering whether we need to duplicate the documentation in both the source code and the website. The only problem is the translation, maybe we can start translating a part of the website.

The download page can contain both the AppImage and the deb package.

I'm also wondering whether it is helpful to directly show the download URL in the update checker. I think it's OK to show the release page. What do you think?

**Comment by coder3101** [comment]
> I tested the CI in my fork and tested the package with [debtap](https://aur.archlinux.org/packages/debtap) on Arch Linux. It works.
> 
> Now the problems are the documentation, download page and the update checker.
> 
> I'm wondering whether we need to duplicate the documentation in both the source code and the website. The only problem is the translation, maybe we can start translating a part of the website, and move links to the website.

Yes, we should focus more on the website and get rid of most of things except changelog and manual. 

> 
> The download page can contain both the AppImage and the deb package.
> 

In the selection we can have Linux AppImage, Linux Deb.


> I'm also wondering whether it is helpful to directly show the download URL in the update checker. I think it's OK to show the release page. What do you think?

On Linux we can show direct download button of both (Download AppImage, Download Deb). We can simply add another button widget in the layout.

**Comment by IZOBRETATEL777** [comment]
> I'm also wondering whether it is helpful to directly show the download URL in the update checker. I think it's OK to show the release page. What do you think?

I think too. If we can do releases on Snapstore or on package repository (Launchpad) there will be no problems with importance of direct links to deb or appimage

**Comment by IZOBRETATEL777** [comment]
We have one small problem with changelog file too. It should be like [this](https://www.debian.org/doc/debian-policy/ch-source.html#s-dpkgchangelog). There is no problems now with using, installing or packaging but it is standard. I think that we will publish .deb releases, but we can apply this standard of keeping changelog in the future.

**Comment by IZOBRETATEL777** [comment]
@coder3101 , @ouuan please tell me if you agree with such changes in INSTALL.md then I'll translate it into Russian

**Comment by coder3101** [comment]
First open a PR to website then for Install.md 

In future we can simply remove install.md from doc/*

After this PR, if you have time. You could help us translate the website in Russian.

**Comment by ouuan** [comment]
> Yes, we should focus more on the website and get rid of most of things except changelog and manual.

I think the manual can be moved to the website, too.

BTW, how long can you own the cpeditor.org domain? I'd rather link to cpeditor.github.io as it is more likely to be available after many years, and now it will be redirected to cpeditor.org.

**Comment by coder3101** [comment]
@ouuan I will own it for many currently its for 1 years. I will renew it next year for 3 or 5 years.

**Comment by ouuan** [comment]
> On Linux we can show direct download button of both (Download AppImage, Download Deb). We can simply add another button widget in the layout.

There could be more types of releases in the future. I think linking to the release page is OK.

**Comment by coder3101** [comment]
Rest assured I will make sure to not lose the domain.

> There could be more types of releases in the future. I think linking to the release page is OK.
That's fine too, but it should only be for Linux. Am I correct?

**Comment by ouuan** [comment]
> That's fine too, but it should only be for Linux. Am I correct?

As for now, there are setup version and portable version for Windows. Now we provide the portable release link for portable version users and vice versa, though. If you think the download link is very convenient, it can be for Linux only.

Another solution is to add a macro when building releases, which identifies the type of the release.

However, if the deb package is uploaded to somewhere that makes it convenient to update, we could only link to the AppImage, because deb users have better ways to update.

**Review by coder3101 — APPROVED** [review]

**Comment by coder3101** [comment]
@ouuan If you have some time. You can review the PR, otherwise or else I will merge it in the next 24 hours.

**Comment by coder3101** [comment]
@all-contributors add @IZOBRETATEL777 for infra

**Review by ouuan — APPROVED** [review]
The package and the update checker are fine.

1.  If you want to release the rpm package, you can do it now.
2.  We'd better choose AppImage as the default in the download page: https://github.com/cpeditor/cpeditor.github.io/blob/a10d1f68fad6e585dcce1c2d11771b5f611b9073/static/js/api/download.js#L94-L96

**Comment by coder3101** [comment]
I think it's good to have a different PR for RPM as this thread is already so long. And our principals (1 PR, 1 feature)

**Pull request merged**
